README.md
# omni_turtlesim

| dashing-devel | eloquent-devel | foxy-devel |
| ---- | ---- | ---- |
| [![industrial_ci (dashing)](https://github.com/Tiryoh/omni_turtlesim_ros2/workflows/industrial_ci%20(dashing)/badge.svg?branch=dashing-devel)](https://github.com/Tiryoh/omni_turtlesim_ros2/actions?query=workflow%3A%22industrial_ci+%28dashing%29%22+branch%3Adashing-devel) | [![industrial_ci (eloquent)](https://github.com/Tiryoh/omni_turtlesim_ros2/workflows/industrial_ci%20(eloquent)/badge.svg?branch=eloquent-devel)](https://github.com/Tiryoh/omni_turtlesim_ros2/actions?query=workflow%3A%22industrial_ci+%28eloquent%29%22+branch%3Aeloquent-devel) | [![industrial_ci (foxy)](https://github.com/Tiryoh/omni_turtlesim_ros2/workflows/industrial_ci%20(foxy)/badge.svg?branch=foxy-devel)](https://github.com/Tiryoh/omni_turtlesim_ros2/actions?query=workflow%3A%22industrial_ci+%28foxy%29%22+branch%3Afoxy-devel) |

ROS 2 package of omni_turtlesim

omni-directional version of [turtlesim](http://wiki.ros.org/turtlesim)

![Image from Gyazo](https://i.gyazo.com/974e67e38431b10c9985c9b033eed577.gif)

## Usage

```bash
# On Terminal 1
$ ros2 run omni_turtlesim turtlesim_node
# On Terminal 2
$ ros2 run teleop_twist_keyboard teleop_twist_keyboard cmd_vel:=/turtle1/cmd_vel
```

## ROS Nodes

### turtlesim_node

turtlesim_node provides a simple simulator.

#### Subscribed Topics

* __turtleX/cmd_vel__ (geometry_msgs/Twist)
    * The linear and angular command velocity for turtleX. The turtle will execute a velocity command for 1 second then time out. Twist.linear.x is the forward velocity and Twist.angular.z is the angular velocity.

#### Published Topics

* __turtleX/pose__ (omni_turtlesim/Pose)
    * The x, y, theta, linear velocity, and angular velocity of turtleX.
* __turtleX/color_sensor__ (omni_turtlesim/Color)
    * The RGB color on the color sensor on turtleX.

#### Services

* __clear__ (std_srvs/Empty)
    * Clears the turtlesim background and sets the color to the value of the background parameters.
* __reset__ (std_srvs/Empty)
    * Resets the turtlesim to the start configuration and sets the background color to the value of the background.
* __kill__ (omni_turtlesim/Kill)
    * Kills a turtle by name.
* __spawn__ (omni_turtlesim/Spawn)
    * Spawns a turtle at (x, y, theta) and returns the name of the turtle. Also will take name for argument but will fail if a duplicate name.
* __turtleX/set_pen__ (omni_turtlesim/SetPen)
    * Sets the pen's color (r g b), width (width), and turns the pen on and off (off).
* __turtleX/teleport_absolute__ (omni_turtlesim/TeleportAbsolute)
    * Teleports the turtleX to (x, y, theta).
* __turtleX/teleport_relative__ (omni_turtlesim/TeleportRelative)
    * Teleports the turtleX a linear and angular distance from the turtles current position.

#### Parameters
* __~background_b__ (int, default: 255)
    * Sets the blue channel of the background color.
* __~background_g__ (int, default: 86)
    * Sets the green channel of the background color.
* __~background_r__ (int, default: 69)
    * Sets the red channel of the background color.

### mimic

mimic provides a simple interface for making one turtlesim mimic another.

#### Subscribed Topics
* __input/pose__ (omni_turtlesim/Pose)
    * The input topic for the mimic node. The topic must be remapped to the pose topic of the desired turtle to mimic.

#### Published Topics
* __output/cmd_vel__ (geometry_msgs/Twist)
    * The output topic for the mimic node. The topic must be remapped to the cmd_vel topic of the mimicking turtle.


## License

(C) 2020 Daisuke Sato

This repository is released under the 3-Clause BSD License ([BSD-3-Clause](https://spdx.org/licenses/BSD-3-Clause.html)), see [LICENSE](./LICENSE).

### Acknowledgements

This software is based on [turtlesim](http://wiki.ros.org/turtlesim), released under the 3-Clause BSD license.

```cpp
/*
 * Copyright (c) 2009, Willow Garage, Inc.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 *     * Redistributions of source code must retain the above copyright
 *       notice, this list of conditions and the following disclaimer.
 *     * Redistributions in binary form must reproduce the above copyright
 *       notice, this list of conditions and the following disclaimer in the
 *       documentation and/or other materials provided with the distribution.
 *     * Neither the name of the Willow Garage, Inc. nor the names of its
 *       contributors may be used to endorse or promote products derived from
 *       this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */
 ```

CHANGELOG.rst
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Changelog for package omni_turtlesim
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

2.2.0 (2020-07-18)
------------------
* docs: Add README and LICENSE (`#17 <https://github.com/Tiryoh/omni_turtlesim_ros2/issues/17>`_)
* ci: Update CI name for logging (`#18 <https://github.com/Tiryoh/omni_turtlesim_ros2/issues/18>`_)
* feat: Update package description (`#16 <https://github.com/Tiryoh/omni_turtlesim_ros2/issues/16>`_)
* fix: Rename package name in launch (`#5 <https://github.com/Tiryoh/omni_turtlesim_ros2/issues/5>`_)
* feat: Add omni control (`#4 <https://github.com/Tiryoh/omni_turtlesim_ros2/issues/4>`_)
* feat: Rename package (`#3 <https://github.com/Tiryoh/omni_turtlesim_ros2/issues/3>`_)
* ci: Add industrial_ci settings
* Contributors: Daisuke Sato, Tiryoh

1.2.4 (2020-06-12)
------------------
* add step value to turtlesim color parameters (`#91 <https://github.com/ros/ros_tutorials/issues/91>`_)

1.2.3 (2020-06-05)
------------------
* update Foxy turtle (`#90 <https://github.com/ros/ros_tutorials/issues/90>`_)
* Contributors: Dirk Thomas

1.2.2 (2020-06-03)
------------------
* add Foxy turtle (`#89 <https://github.com/ros/ros_tutorials/issues/89>`_)
* Contributors: Jacob Perron

1.2.1 (2020-06-02)
------------------
* Fix Qt deprecation warning (`#88 <https://github.com/ros/ros_tutorials/issues/88>`_)
* Contributors: Jacob Perron

1.2.0 (2020-05-01)
------------------
* Replace deprecated launch_ros usage (`#84 <https://github.com/ros/ros_tutorials/issues/84>`_)
* Contributors: Jacob Perron

1.1.1 (2020-04-16)
------------------
* catch reference to fix -Wcatch-value warning (`#78 <https://github.com/ros/ros_tutorials/issues/78>`_)
* Contributors: Mikael Arguedas

1.1.0 (2019-11-12)
------------------
* Eloquent Elusor turtle icon (`#77 <https://github.com/ros/ros_tutorials/issues/77>`_)

1.0.2 (2019-10-23)
------------------
* separate background color from drawn paths, trigger repaint on parameter changes (`#75 <https://github.com/ros/ros_tutorials/issues/75>`_)
* add descriptor information for background color parameters (`#73 <https://github.com/ros/ros_tutorials/issues/73>`_)
* Fix Windows compiler warning (`#69 <https://github.com/ros/ros_tutorials/issues/69>`_)
* Change log messages to use 'goal' instead of 'action' (`#67 <https://github.com/ros/ros_tutorials/issues/67>`_)

1.0.1 (2019-10-02)
------------------
* fix mimic tutorial node (`#65 <https://github.com/ros/ros_tutorials/issues/65>`_)
* fix syntax error in teleop_turtle_key.cpp on Windows (`#66 <https://github.com/ros/ros_tutorials/issues/66>`_)
* add RotateAbsolute action (`#62 <https://github.com/ros/ros_tutorials/issues/62>`_)
* fix typo in error message (`#64 <https://github.com/ros/ros_tutorials/issues/64>`_)

1.0.0 (2019-09-24)
------------------
* replace images with ROS 2 turtles (`#60 <https://github.com/ros/ros_tutorials/issues/60>`_)
* add shortcut to quit teleop (`#58 <https://github.com/ros/ros_tutorials/issues/58>`_)
* fix compiler warnings on Windows (`#57 <https://github.com/ros/ros_tutorials/issues/57>`_)
* add support for Windows (`#56 <https://github.com/ros/ros_tutorials/issues/56>`_)
* various fixes for ROS 2 (`#55 <https://github.com/ros/ros_tutorials/issues/55>`_)
* turtlesim for ROS 2 (`#53 <https://github.com/ros/ros_tutorials/issues/53>`_)

0.9.1 (2019-03-04)
------------------
* change formula to avoid rounding with extreme input values (`#51 <https://github.com/ros/ros_tutorials/issues/51>`_)
* keep theta in the desired interval (`#46 <https://github.com/ros/ros_tutorials/issues/46>`_)

0.9.0 (2018-04-11)
------------------
* add melodic turtle (`#41 <https://github.com/ros/ros_tutorials/issues/41>`_)

0.8.1 (2017-07-27)
------------------
* theta ranges from -pi to +pi (`#31 <https://github.com/ros/ros_tutorials/issues/31>`_)

0.8.0 (2017-03-10)
------------------
* add lunar turtle (`#39 <https://github.com/ros/ros_tutorials/pull/39>`_)

0.7.1 (2016-10-24)
------------------
* check pen_on\_ when processing teleport requests (`#35 <https://github.com/ros/ros_tutorials/pull/35>`_)

0.7.0 (2016-03-18)
------------------
* add kinetic image
* update to Qt5
* fix size of Jade image to not exceed other images in order to not get positioned incorrectly
* fix compiler warnings

0.6.1 (2015-09-19)
------------------
* update the coordinate system in /spawn service for consistency (`#25 <https://github.com/ros/ros_tutorials/pull/25>`_)

0.6.0 (2015-05-21)
------------------
* add jade turtle (`#22 <https://github.com/ros/ros_tutorials/pull/22>`_)

0.5.3 (2015-05-04)
------------------

0.5.2 (2014-12-23)
------------------

0.5.1 (2014-05-08)
------------------

0.5.0 (2014-05-07)
------------------
* add indigo turtle
* add disabled code to easily spawn all available turtle types

0.4.3 (2014-01-07)
------------------

0.4.2 (2013-10-04)
------------------
* fix missing install of hydro.svg (`#12 <https://github.com/ros/ros_tutorials/issues/12>`_)

0.4.1 (2013-09-11)
------------------
* add hydro image to turtlesim

0.4.0 (2013-09-06)
------------------
* Adding png version of hydro for wiki linking
* TurtleApp accepts argc by reference
* Restoring all the changes appropriate for Hydro

0.3.13 (2013-08-21)
-------------------
* TurtleApp accepts argc by reference
* add hydro image to turtlesim
* remove mainpage.dox

0.3.12 (2013-03-29)
-------------------
* reverting velocity -> twist for groovy
* Revert "chaning command_velocity to cmd_vel" for groovy
  This reverts commit 96e5174d3a5c961b6e1195b90b4024e2858df010.
* Revert "adding geometry_msgs dependency in package.xml and CMakelist" for groovy
  This reverts commit c7ac1155d70269909b55af03d13fe2e089d6215d.
* Revert "alaphabetic order" for groovy
  This reverts commit f928765ed08773517c195b74c55231c0e4fcc5e5.

0.3.11 (2013-03-21)
-------------------
* update email in package.xml

0.3.10 (2013-03-08)
-------------------
* Fix a moc generation error with boost >= 1.48
  See:

src/turtle_frame.cpp
/*
 * Copyright (c) 2009, Willow Garage, Inc.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 *     * Redistributions of source code must retain the above copyright
 *       notice, this list of conditions and the following disclaimer.
 *     * Redistributions in binary form must reproduce the above copyright
 *       notice, this list of conditions and the following disclaimer in the
 *       documentation and/or other materials provided with the distribution.
 *     * Neither the name of the Willow Garage, Inc. nor the names of its
 *       contributors may be used to endorse or promote products derived from
 *       this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

#include "turtlesim/turtle_frame.h"

#include <QPointF>

#include <cstdlib>
#include <ctime>

#define DEFAULT_BG_R 0x45
#define DEFAULT_BG_G 0x56
#define DEFAULT_BG_B 0xff

namespace omni_turtlesim
{

TurtleFrame::TurtleFrame(rclcpp::Node::SharedPtr& node_handle, QWidget* parent, Qt::WindowFlags f)
: QFrame(parent, f)
, path_image_(500, 500, QImage::Format_ARGB32)
, path_painter_(&path_image_)
, frame_count_(0)
, id_counter_(0)
{
  setFixedSize(500, 500);
  setWindowTitle("TurtleSim");

  srand(time(NULL));

  update_timer_ = new QTimer(this);
  update_timer_->setInterval(16);
  update_timer_->start();

  connect(update_timer_, SIGNAL(timeout()), this, SLOT(onUpdate()));

  nh_ = node_handle;
  rcl_interfaces::msg::IntegerRange range;
  range.from_value = 0;
  range.step = 1;
  range.to_value = 255;
  rcl_interfaces::msg::ParameterDescriptor background_r_descriptor;
  background_r_descriptor.description = "Red channel of the background color";
  background_r_descriptor.integer_range.push_back(range);
  rcl_interfaces::msg::ParameterDescriptor background_g_descriptor;
  background_g_descriptor.description = "Green channel of the background color";
  background_g_descriptor.integer_range.push_back(range);
  rcl_interfaces::msg::ParameterDescriptor background_b_descriptor;
  background_b_descriptor.description = "Blue channel of the background color";
  background_b_descriptor.integer_range.push_back(range);
  nh_->declare_parameter("background_r", rclcpp::ParameterValue(DEFAULT_BG_R), background_r_descriptor);
  nh_->declare_parameter("background_g", rclcpp::ParameterValue(DEFAULT_BG_G), background_g_descriptor);
  nh_->declare_parameter("background_b", rclcpp::ParameterValue(DEFAULT_BG_B), background_b_descriptor);

  QVector<QString> turtles;
  turtles.append("ardent.png");
  turtles.append("bouncy.png");
  turtles.append("crystal.png");
  turtles.append("dashing.png");
  turtles.append("eloquent.png");
  turtles.append("foxy.png");

  QString images_path = (ament_index_cpp::get_package_share_directory("omni_turtlesim") + "/images/").c_str();
  for (int i = 0; i < turtles.size(); ++i)
  {
    QImage img;
    img.load(images_path + turtles[i]);
    turtle_images_.append(img);
  }

  meter_ = turtle_images_[0].height();

  clear();

  clear_srv_ = nh_->create_service<std_srvs::srv::Empty>("clear", std::bind(&TurtleFrame::clearCallback, this, std::placeholders::_1, std::placeholders::_2));
  reset_srv_ = nh_->create_service<std_srvs::srv::Empty>("reset", std::bind(&TurtleFrame::resetCallback, this, std::placeholders::_1, std::placeholders::_2));
  spawn_srv_ = nh_->create_service<omni_turtlesim::srv::Spawn>("spawn", std::bind(&TurtleFrame::spawnCallback, this, std::placeholders::_1, std::placeholders::_2));
  kill_srv_ = nh_->create_service<omni_turtlesim::srv::Kill>("kill", std::bind(&TurtleFrame::killCallback, this, std::placeholders::_1, std::placeholders::_2));

  rclcpp::QoS qos(rclcpp::KeepLast(100), rmw_qos_profile_sensor_data);
  parameter_event_sub_ = nh_->create_subscription<rcl_interfaces::msg::ParameterEvent>(
    "/parameter_events", qos, std::bind(&TurtleFrame::parameterEventCallback, this, std::placeholders::_1));

  RCLCPP_INFO(nh_->get_logger(), "Starting turtlesim with node name %s", nh_->get_node_names()[0].c_str());

  width_in_meters_ = (width() - 1) / meter_;
  height_in_meters_ = (height() - 1) / meter_;
  spawnTurtle("", width_in_meters_ / 2.0, height_in_meters_ / 2.0, 0);

  // spawn all available turtle types
  if(false)
  {
    for(int index = 0; index < turtles.size(); ++index)
    {
      QString name = turtles[index];
      name = name.split(".").first();
      name.replace(QString("-"), QString(""));
      spawnTurtle(name.toStdString(), 1.0f + 1.5f * (index % 7), 1.0f + 1.5f * (index / 7), static_cast<float>(PI) / 2.0f, index);
    }
  }
}

TurtleFrame::~TurtleFrame()
{
  delete update_timer_;
}

bool TurtleFrame::spawnCallback(const omni_turtlesim::srv::Spawn::Request::SharedPtr req, omni_turtlesim::srv::Spawn::Response::SharedPtr res)
{
  std::string name = spawnTurtle(req->name, req->x, req->y, req->theta);
  if (name.empty())
  {
    RCLCPP_ERROR(nh_->get_logger(), "A turtle named [%s] already exists", req->name.c_str());
    return false;
  }

  res->name = name;

  return true;
}

bool TurtleFrame::killCallback(const omni_turtlesim::srv::Kill::Request::SharedPtr req, omni_turtlesim::srv::Kill::Response::SharedPtr)
{
  M_Turtle::iterator it = turtles_.find(req->name);
  if (it == turtles_.end())
  {
    RCLCPP_ERROR(nh_->get_logger(), "Tried to kill turtle [%s], which does not exist", req->name.c_str());
    return false;
  }

  turtles_.erase(it);
  update();

  return true;
}

void TurtleFrame::parameterEventCallback(const rcl_interfaces::msg::ParameterEvent::SharedPtr event)
{
  // only consider events from this node
  if (event->node == nh_->get_fully_qualified_name())
  {
    // since parameter events for this even aren't expected frequently just always call update()
    update();
  }
}

bool TurtleFrame::hasTurtle(const std::string& name)
{
  return turtles_.find(name) != turtles_.end();
}

std::string TurtleFrame::spawnTurtle(const std::string& name, float x, float y, float angle)
{
  return spawnTurtle(name, x, y, angle, rand() % turtle_images_.size());
}

std::string TurtleFrame::spawnTurtle(const std::string& name, float x, float y, float angle, size_t index)
{
  std::string real_name = name;
  if (real_name.empty())
  {
    do
    {
      std::stringstream ss;
      ss << "turtle" << ++id_counter_;
      real_name = ss.str();
    } while (hasTurtle(real_name));
  }
  else
  {
    if (hasTurtle(real_name))
    {
      return "";
    }
  }

  TurtlePtr t = std::make_shared<Turtle>(nh_, real_name, turtle_images_[static_cast<int>(index)], QPointF(x, height_in_meters_ - y), angle);
  turtles_[real_name] = t;
  update();

  RCLCPP_INFO(nh_->get_logger(), "Spawning turtle [%s] at x=[%f], y=[%f], theta=[%f]", real_name.c_str(), x, y, angle);

  return real_name;
}

void TurtleFrame::clear()
{
  // make all pixels fully transparent
  path_image_.fill(qRgba(255, 255, 255, 0));
  update();
}

void TurtleFrame::onUpdate()
{
  if (!rclcpp::ok())
  {
    close();
    return;
  }

  rclcpp::spin_some(nh_);

  updateTurtles();
}

void TurtleFrame::paintEvent(QPaintEvent*)
{
  QPainter painter(this);

  int r = DEFAULT_BG_R;
  int g = DEFAULT_BG_G;
  int b = DEFAULT_BG_B;
  nh_->get_parameter("background_r", r);
  nh_->get_parameter("background_g", g);
  nh_->get_parameter("background_b", b);
  QRgb background_color = qRgb(r, g, b);
  painter.fillRect(0, 0, width(), height(), background_color);

  painter.drawImage(QPoint(0, 0), path_image_);

  M_Turtle::iterator it = turtles_.begin();
  M_Turtle::iterator end = turtles_.end();
  for (; it != end; ++it)
  {
    it->second->paint(painter);
  }
}

void TurtleFrame::updateTurtles()
{
  if (last_turtle_update_.nanoseconds() == 0)
  {
    last_turtle_update_ = nh_->now();
    return;
  }

  bool modified = false;
  M_Turtle::iterator it = turtles_.begin();
  M_Turtle::iterator end = turtles_.end();
  for (; it != end; ++it)
  {
    modified |= it->second->update(0.001 * update_timer_->interval(), path_painter_, path_image_, width_in_meters_, height_in_meters_);
  }
  if (modified)
  {
    update();
  }

  ++frame_count_;
}


bool TurtleFrame::clearCallback(const std_srvs::srv::Empty::Request::SharedPtr, std_srvs::srv::Empty::Response::SharedPtr)
{
  RCLCPP_INFO(nh_->get_logger(), "Clearing turtlesim.");
  clear();
  return true;
}

bool TurtleFrame::resetCallback(const std_srvs::srv::Empty::Request::SharedPtr, std_srvs::srv::Empty::Response::SharedPtr)
{
  RCLCPP_INFO(nh_->get_logger(), "Resetting turtlesim.");
  turtles_.clear();
  id_counter_ = 0;
  spawnTurtle("", width_in_meters_ / 2.0, height_in_meters_ / 2.0, 0);
  clear();
  return true;
}

}
